src/Microsoft.OData.Core/Uri/ODataUriSlim.cs

```diff
@@ -0,0 +1,139 @@
+﻿// <copyright file="ODataUriSlim.cs" company="Microsoft">
+//      Copyright (C) Microsoft Corporation. All rights reserved. See License.txt in the project root for license information.
+// </copyright>
+//---------------------------------------------------------------------
+
+using System;
+using System.Diagnostics;
+using System.Collections.Generic;
+using Microsoft.OData.UriParser.Aggregation;
+using Microsoft.OData.UriParser;
+
+namespace Microsoft.OData
+{
+    /// <summary>
+    /// A lightweight version of <see cref="ODataUri"/> that is
+    /// used by the writer to efficiently create a copy of the URI when entering
+    /// each scope. This is used to avoid cloning <see cref="ODataUri"/>
+    /// excessively.
+    /// </summary>
+    internal struct ODataUriSlim

```

Comment:
can we 'implicitly' convert between ODataUri and ODatUriSlim, back and forth?

Reply:
We could add implicit conversion from `ODataUri` to `ODataUriSlim` such that someone is able to do assign an `ODataUriSlim` variable an `ODataUri` value:
```c#
ODataUri odataUri = ...;
ODataUriSlim odataUriSlim = odataUri;
```
or pass an `ODataUri` to a method expecting an `ODataUriSlim`:

```c#
public void Foo(in ODataUriSlim odataUriSlim)
{
 ....
}
// ..

ODataUri odataUri ...;
Foo(odataUri)
```

But I don't think we should add conversion from `ODataUriSlim` to `ODataUri`. Since the `ODataUri` instance referenced in `ODataUriSlim` might have different values for some of the properties, exposing it could lead to inconsistencies or unwanted mutations to the `ODataUri`. To avoid that, we may need to clone the `ODataUri`, which beats the purpose of `ODataUriSlim`.

In the current code, I don't see a lot of use-cases for `ODataUri`-`ODataUriSlim` implicit conversions. I also think keeping things explicit (like in the current PR) makes it less likely that someone will use the wrong type accidentally.